Enhance YouTube transcript retrieval logic

akhilender-bongirwar opened the pull request:
fixes #245 
/claim #245 

### Summary
This PR improves the logic for handling YouTube transcript retrieval within the application. Key changes include:
- Updated the tool invocation handler to check if the YouTube transcript is successfully retrieved based on the selected videos.
- Improved the message filtering logic to exclude assistant messages that indicate a transcript retrieval issue if the transcript was successfully retrieved.
- Added state management for transcript retrieval status to enhance user experience and feedback.
- Ensured that the component correctly displays success or error messages based on the state of transcript retrieval.

## Testing

https://github.com/user-attachments/assets/311c1447-031d-4232-815f-5285d8dc2e29

In the video demonstration, third video example does not have transcripts provided by Youtube. Hence throws error as expected.

aexshafii commented:
@akhilender-bongirwar 

### Critique of the Pull Request (PR)

The PR aims to address an issue in the File Organizer 2000 Obsidian plugin where the AI assistant incorrectly states that it couldn't retrieve a YouTube transcript, even when the transcript was successfully fetched. The current approach attempts to filter out such erroneous messages on the frontend.

---

### **Analysis**

#### Strengths:
- **Identifies the Problem:** The PR correctly identifies that unwanted error messages are displayed to the user even when the transcript retrieval was successful.
- **Attempts a Fix:** By filtering messages and excluding certain assistant messages based on conditions, it tries to prevent the incorrect error message from appearing.

#### Weaknesses:
- **Frontend Filtering is Fragile:**
  - **Race Conditions:** The state `transcriptRetrieved` may not update synchronously with the arrival of messages, leading to inconsistent filtering.
  - **Incomplete Filtering:** The assistant's message may contain valuable information alongside the error message, which could be improperly filtered out.
  
- **Complexity in Message Handling:**
  - Filtering messages based on content strings and roles adds unnecessary complexity to the component.
  - The approach relies on hardcoded error messages, which is not scalable or maintainable.

- **Root Cause Not Addressed:** 
  - The assistant should not generate incorrect error messages in the first place. The PR addresses the symptom (the message) rather than the root cause (the assistant's logic).

---

### **Recommendations for Improvement**

1. **Handle Errors at the Source**
   - **Modify the Tool Invocation Logic:**
     - Ensure that the assistant only mentions the inability to retrieve the transcript when the retrieval actually fails.
     - **a. Update `handleToolCall` Function**
       ```tsx
if (toolCall.toolName === "getYouTubeTranscript") {
  const args = toolCall.args as { videoId: string };
  const { videoId } = args;
  try {
    const transcript = await getYouTubeTranscript(videoId);
    const title = await getYouTubeVideoTitle(videoId);
    setSelectedYouTubeVideos(prev => [
      ...prev,
      { videoId, title, transcript },
    ]);
    addToolResult({
      toolCallId: toolCall.toolCallId,
      result: { transcript, title, videoId },
    });
    // Indicate success explicitly
    return { success: true };
  } catch (error) {
    console.error("Error fetching YouTube transcript:", error);
    // Return error details to the assistant
    return { success: false, error: error.message };
  }
}
       ```

     - **b. Adjust Assistant's Response Logic**
       ```tsx
// Pseudo-code representation of assistant logic
if (toolResult.success) {
  // Proceed with using the transcript
  assistantMessage = `I've successfully retrieved the transcript for the video titled "${title}". How can I assist you further with it?`;
} else {
  // Inform the user of the failure
  assistantMessage = `I couldn't retrieve the transcript for the YouTube video with the ID "${videoId}". ${toolResult.error}`;
}
       ```

2. **Remove Frontend Message Filtering**
   - Since the assistant will now only generate appropriate messages, you can eliminate the frontend filtering logic. This reduces complexity and the potential for bugs related to state management and message synchronization.
   - **Remove the Following Code:**
     ```tsx
const errorMessages = [
  "don't have access to the transcript for the YouTube video",
  "couldn't retrieve the transcript for the YouTube video",
];

// ...

const filteredMessages = messages.filter(message => {
  const isAssistantMessage =
    message.role === "assistant" && message.content.includes("transcript");

  const isErrorMessage = errorMessages.some(errorMessage =>
    message.content.includes(errorMessage)
  );

  return !(isAssistantMessage && isErrorMessage && transcriptRetrieved);
});
     ```

   - **Update the message rendering to use messages directly:**
     ```tsx
{messages.map(message => (
  // ... render messages without filtering
))}
     ```

3. **Improve Assistant's Prompt or Backend Logic**
   - If using a backend AI service or language model, adjust the prompt engineering or the backend logic to handle the success and error cases more effectively.
   - **Example Prompt Adjustment:**
     - **Before:**
       ```
       Whenever you attempt to retrieve a YouTube transcript, inform the user if you cannot access it.
       ```

     - **After:**
       ```
       When you attempt to retrieve a YouTube transcript, only inform the user if the retrieval fails. If successful, proceed without mentioning any errors.
       ```

4. **Provide Clear User Feedback**
   - If the transcript is successfully retrieved, consider informing the user explicitly. This enhances the user experience by confirming that their request was fulfilled.
   - **Example Assistant Message Upon Success:**
     ```
     I've successfully retrieved the transcript for the video titled "Your Video Title". How can I assist you further with it?
     ```

---

### **Is There a Better Approach?**
Yes, handling the error at the source (during the tool invocation) and ensuring the assistant constructs messages based on the actual outcome is a better approach. This avoids the need for frontend message filtering, reduces complexity, and provides a cleaner, more maintainable codebase.

---

### **Conclusion**

- **Current PR Limitations:** While the PR attempts to fix the issue, it doesn't address the root cause and introduces complexity through frontend filtering.
- **Recommended Solution:** Handle success and failure cases within the tool invocation logic and adjust the assistant's response accordingly. Remove unnecessary frontend filtering.

---

### **Benefits**

- **Reliability:** Ensures that messages accurately reflect the outcome of transcript retrieval.
- **Maintainability:** Reduces code complexity and potential bugs.
- **User Experience:** Provides clear and accurate feedback to the user.

---

### **Improved Code Snippets**

- **In `chat.tsx`:**
   ```tsx
// Remove error messages array and filtering logic
// const errorMessages = [...];

// Remove filteredMessages and use messages directly in rendering
// const filteredMessages = ...

// In handleToolCall function
const handleToolCall = async ({ toolCall }) => {
  if (toolCall.toolName === "getYouTubeTranscript") {
    const args = toolCall.args as { videoId: string };
    const { videoId } = args;
    try {
      const transcript = await getYouTubeTranscript(videoId);
      const title = await getYouTubeVideoTitle(videoId);
      setSelectedYouTubeVideos(prev => [
        ...prev,
        { videoId, title, transcript },
      ]);
      addToolResult({
        toolCallId: toolCall.toolCallId,
        result: { transcript, title, videoId },
      });
      // Return success flag
      return { success: true, transcriptRetrieved: true, title };
    } catch (error) {
      console.error("Error fetching YouTube transcript:", error);
      return { success: false, transcriptRetrieved: false, error: error.message };
    }
  }
  // ... handle other tools
};
   ```

- **In Assistant Logic or Prompt:**
   ```tsx
   // Ensure that the assistant constructs its response based on the success flag.
   ```

---

### **Summary**
By handling success and failure cases explicitly during the tool invocation and adjusting the assistant's response logic, we can ensure accurate communication with the user without relying on fragile frontend filtering. This results in a more robust and maintainable solution.

aexshafii commented:
@akhilender-bongirwar thanks for the PR. This approach does however not work consistently. Here are some suggestions on how to implement a better solution:

![Screenshot 2024-10-07 at 1 06 16 PM](https://github.com/user-attachments/assets/a8d67993-3f8b-42f2-b15a-853de0d5dcc4)
![Screenshot 2024-10-07 at 1 06 37 PM](https://github.com/user-attachments/assets/be5b03dd-bdc5-4bdd-94f9-82520ae0ad78)
![Screenshot 2024-10-07 at 1 06 50 PM](https://github.com/user-attachments/assets/da09df36-cc9a-4b16-aded-a357034a87ae)

akhilender-bongirwar commented:
@aexshafii,  Thank you for your detailed feedback and suggestions. I understand your concerns regarding frontend filtering, and I agree that addressing issues at the source is typically a better approach. However, in this particular case, I chose to implement the filtering logic on the frontend for a specific reason:

Whenever I provide a YouTube video link and request a transcript, the response from the OpenAI API results from the assistant: one initiating the tool call and the other an error message indicating that the transcript could not be retrieved (shown in image below). This happens consistently, even when the transcript is successfully fetched. The error message is part of the assistant's response, which makes it difficult to prevent from being displayed on the frontend.

Since I can't control the response generation from the OpenAI API (using `useChat` from the `ai-sdk`), I realized that filtering these messages based on the context where the transcript is stored was the most reliable approach. The logic ensures that if the transcript is successfully retrieved and stored in the context, I can confidently filter out the erroneous error message. This approach guarantees that only valid messages are displayed, without impacting any further interactions regarding the video.

I acknowledge that frontend filtering adds a layer of complexity, and it's not an ideal long-term solution. However, given the constraints where I cannot modify the underlying API behavior, this solution effectively prevents the erroneous error message from being shown while maintaining the integrity of the information stored in the context. Even if I were to implement `success flags` as you suggested, they would still be dependent on the AI's response, which includes the error message by default. As a result, I would still need to filter the messages based on the content.

![image](https://github.com/user-attachments/assets/834e82bf-1241-467d-b291-ca8653f0b953)

aexshafii commented:
@akhilender-bongirwar thanks for your help with this issue but for now I estimated the other solution was satisfactory and introduced minimal changes. I will make sure to reward you with the bounty if we do implement this solution later on 🙏

akhilender-bongirwar commented:
> @akhilender-bongirwar thanks for your help with this issue but for now I estimated the other solution was satisfactory and introduced minimal changes. I will make sure to reward you with the bounty if we do implement this solution later on 🙏


@aexshafii, Thank you for your consideration, but these changes correctly ensure the successful retrieval of YouTube transcripts.

akhilender-bongirwar commented:
@aexshafii, I just wanted to follow up, as it's been a week since I made the changes. I was wondering if you’ve had a chance to test them and whether you've encountered any issues. I made sure everything aligns with what’s shown on [fileorganizer2000.com](https://fileorganizer2000.com/) for retrieving YouTube transcripts and interacting with the content.

![image](https://github.com/user-attachments/assets/6b8a1616-edc9-4a3c-9bff-2dc454ca837a)

Thanks so much for your time, and I appreciate your feedback.
